### PR TITLE
[Core] Optimize kubernetes cmd executions with kubernetes command runner

### LINKED ISSUE #3154: [k8s] Slow `file_mounts` on k8s
Opening a SSH connection on k8s is high latency because of two extra steps:

1. Goes through k8s API server (`kubectl port-forward`)
2. Jumps through SSH jump pod.

This makes many operations slow, including file_mounts. Example, mounting three 1kb files takes 22s on EKS K8s compared to 6s on GCP:

```
file_mounts:
  ~/.gitconfig: ~/.gitconfig
  ~/tmp.txt: ~/tmp.txt
  ~/iam_policy.json: ~/iam_policy.json
```

Kubernetes:
```
I 02-13 10:33:42 backend_utils.py:1268] Syncing (to 1 node): ~/.gitconfig -> ~/.gitconfig
D 02-13 10:33:42 command_runner.py:110] --- Proxy: ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  ---
D 02-13 10:33:46 command_runner.py:110] --- Proxy: ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  ---
I 02-13 10:33:49 backend_utils.py:1268] Syncing (to 1 node): ~/tmp.txt -> ~/tmp.txt
D 02-13 10:33:49 command_runner.py:110] --- Proxy: ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  ---
D 02-13 10:33:54 command_runner.py:110] --- Proxy: ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  ---
I 02-13 10:33:57 backend_utils.py:1268] Syncing (to 1 node): ~/iam_policy.json -> ~/iam_policy.json
D 02-13 10:33:57 command_runner.py:110] --- Proxy: ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  ---
D 02-13 10:34:01 command_runner.py:110] --- Proxy: ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  ---
D 02-13 10:34:05 cloud_vm_ray_backend.py:4571] File mount sync took 22.776047945022583 seconds.
```

GCP:
```
I 02-13 10:38:11 backend_utils.py:1268] Syncing (to 1 node): ~/.gitconfig -> ~/.gitconfig
I 02-13 10:38:13 backend_utils.py:1268] Syncing (to 1 node): ~/tmp.txt -> ~/tmp.txt
I 02-13 10:38:15 backend_utils.py:1268] Syncing (to 1 node): ~/iam_policy.json -> ~/iam_policy.json
D 02-13 10:38:18 cloud_vm_ray_backend.py:4571] File mount sync took 6.378494024276733 seconds.
```

Note that it is still fairly high throughput, just that the latency of establishing the connection is high.

Possible fixes:
1. Change our file_mounts codepath for k8s to use `kubectl cp` instead of rsync (and do so in a generic manner which works for other clouds, e.g. SLURM, too)
2. Consolidate all file_mounts into one SSH connection. This is generally good for perf on other clouds too.

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #3154, by refactoring the command runner and avoid using ssh in the backend for kubernetes pods. This is helpful for the future support of Slurm and other services that do not have ssh enabled.

This is also a prototype to reduce the launch and exec time on kubernetes cluster.

We will separate the optimization in this PR to multiple separate PRs:
1. #3394: lazy import and avoid rsync for exec **(1.3x faster for exec)**
2. #3484: avoid source bashrc for internal remote command execution **(1.3x faster for exec)**
3. This PR: kubernetes command runner

With the three optimization we can get **3x** speed up for exec

<details>

<summary>Exec speed (1.8x speed up)</summary>

```
multitime -n 5 sky exec test-speed echo hi -d
```
Master:
```
            Mean        Std.Dev.    Min         Median      Max
real        10.377      0.379       9.948       10.372      10.821      
user        1.373       0.025       1.327       1.381       1.401       
sys         0.178       0.021       0.150       0.171       0.209 
```
This PR
```
            Mean        Std.Dev.    Min         Median      Max
real        5.715       0.146       5.499       5.668       5.925       
user        1.582       0.132       1.475       1.512       1.835       
sys         0.209       0.027       0.170       0.202       0.253  
```

</details>

<details>

<summary>Launch on existing cluster (1.7x speed up)</summary>

`sky launch -c test-existing`
``multitime -n 5 sky launch -y -c test-existing echo hi``
master: 
```
            Mean        Std.Dev.    Min         Median      Max
real        40.467      2.363       37.563      39.865      44.683      
user        3.246       0.573       2.918       2.986       4.389       
sys         0.359       0.065       0.313       0.322       0.487 
```
This PR:
```
            Mean        Std.Dev.    Min         Median      Max
real        23.459      0.304       23.020      23.426      23.918      
user        3.220       0.041       3.155       3.223       3.266       
sys         0.433       0.028       0.389       0.438       0.463 
```

</details>

<details>

<summary>New clusters without setup/run (1.3x speed up)</summary>

``multitime -n 5 sky launch -y --cloud kubernetes --cpus 2``
Master: 
```
            Mean        Std.Dev.    Min         Median      Max
real        55.359      2.039       53.175      54.318      58.499      
user        3.180       0.046       3.112       3.177       3.236       
sys         0.377       0.014       0.356       0.385       0.391  
```
This PR:
```
            Mean        Std.Dev.    Min         Median      Max
real        41.574      1.206       39.681      41.475      43.460      
user        3.809       0.506       3.466       3.577       4.816       
sys         0.553       0.081       0.462       0.540       0.699 
```

</details>

<details>

<summary>New clusters with setup/run (1.4x speed up)</summary>

``multitime -n 5 sky launch -y --cloud kubernetes --cpus 2 task.yaml``
```
setup: echo Some "setup commands"
run: echo Some "run commands"
```
Master:
```
            Mean        Std.Dev.    Min         Median      Max
real        74.712      2.938       71.890      73.015      79.771      
user        3.598       0.465       3.318       3.373       4.526       
sys         0.463       0.104       0.376       0.435       0.661 
```
This PR:
```
            Mean        Std.Dev.    Min         Median      Max
real        54.095      1.973       51.659      54.739      56.846      
user        4.221       0.068       4.174       4.181       4.354       
sys         0.647       0.026       0.599       0.652       0.672 
```

</details>

<details>

<summary>New clusters with file mounts(1.6x speed up)</summary>

`task.yaml`
```yaml
resources:
  cpus: 1
  # image_id: docker:michaelvll/no-local-bin-test:v3
  # image_id: docker:michaelvll/k8s-dep-test:v2

file_mounts:
  /sky/examples: ./examples
  ~/.viminfo : ~/.viminfo
  ~/task.yaml : ./task.yaml
  ~/.config/gcloud: ~/.config/gcloud

run: |
  echo hi
```
``multitime -n 5 sky launch -y --cloud kubernetes --cpus 1 task.yaml``
Master: 
```
            Mean        Std.Dev.    Min         Median      Max
real        99.381      2.770       95.061      101.003     102.193     
user        4.529       0.458       4.255       4.319       5.443       
sys         0.903       0.115       0.821       0.840       1.126  
```
This PR:
```
            Mean        Std.Dev.    Min         Median      Max
real        62.851      2.610       59.383      62.714      67.189      
user        5.895       0.590       5.473       5.558       7.043       
sys         1.366       0.236       1.044       1.333       1.778 
```

</details>

~~Blocked by #3037~~.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c test-ssh --cloud gcp --cpus 2 --num-nodes 2 task.yaml`
  - [x] `sky launch -c test-k8s --cloud kubernetes --cpus 2 --num-nodes 2 task.yaml`
  - [x] `sky launch --cloud aws --cpus 2 task.yaml` with internal ips and jump server.
  - [x] `sky launch --cloud kubernetes --cpus 2 task.yaml`
  - [x]  `sky launch --cloud gcp --cpus 2 task.yaml`
  - [x] `sky launch -c test-docker --cloud gcp --cpus 2 --image-id docker:ubuntu:18.04 task.yaml`
  - [x] `sky launch -c test-docker --cloud aws --cpus 2 --image-id docker:ubuntu:18.04 task.yaml`
    ```yaml
    resources:
      cpus: 2
    
    
    file_mounts:
      /my-examples: ./examples
      ~/task.yaml: ./task.yaml
    
    setup: |
      ls /my-examples
    
    run: |
      ls ~
    ```
- [x] All smoke tests: `pytest tests/test_smoke.py --kubernetes` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
    Make sure `sky logs --sync-down` works for other clouds.
  - [x] `pytest tests/test_smoke.py::test_minimal --aws`
  - [x] `pytest tests/test_smoke.py::test_minimal --gcp`
  - [x] `pytest tests/test_smoke.py::test_minimal --kubernetes`
  - [x] `sky launch --cloud aws --num-nodes 4 -c min echo hi; sky logs --sync-down min` and check the local log dir
- [x] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh` (with stop commands commented out)

### CODE REVIEW

File: sky/utils/command_runner.py [hunk 1]
```diff
@@ -584,3 +584,249 @@ def rsync(
...
+            if rc != 0:
+                raise ValueError('Failed to get remote home directory.')
+            remote_home_dir = remote_home_dir.strip()
+            return remote_home_dir
+
+        # Build command.
+        # TODO(zhwu): This will print a per-file progress bar (with -P),
+        # shooting a lot of messages to the output. --info=progress2 is used
+        # to get a total progress bar, but it requires rsync>=3.1.0 and Mac
+        # OS has a default rsync==2.6.9 (16 years old).
```
Comment (cblmemo) [inline sky/utils/command_runner.py:760]: should we add rsync version requirements to k8s cloud as well, like netcat?

File: sky/utils/command_runner.py [hunk 2]
```diff
@@ -584,3 +584,249 @@ def rsync(
...
+                SkyPilot but we still want to get rid of some warning messages,
+                such as SSH warnings.
+
+        Returns:
+            returncode
+            or
+            A tuple of (returncode, stdout, stderr).
+        """
+        assert port_forward is None, ('port_forward is not supported for k8s '
+                                      f'for now, but got: {port_forward}')
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/utils/command_runner.py:695]: This is possible to be supported by k8s, just not implemented yet. Added a TODO here.

File: sky/utils/command_runner.py [hunk 3]
```diff
@@ -584,3 +584,249 @@ def rsync(
...
+                                    log_path,
+                                    require_outputs=require_outputs,
+                                    stream_logs=stream_logs,
+                                    process_stream=process_stream,
+                                    shell=True,
+                                    executable=executable,
+                                    **kwargs)
+
+    @timeline.event
+    def rsync(
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/utils/command_runner.py:756]: Good catch! Just refactored the code : )

File: sky/utils/command_runner.py [hunk 4]
```diff
@@ -207,6 +212,92 @@ def _get_command_to_run(
...
+            target: str,
+            node_destination: str,
+            up: bool,
+            rsh_option: str,
+            # Advanced options.
+            log_path: str = os.devnull,
+            stream_logs: bool = True,
+            max_retry: int = 1,
+            prefix_command: Optional[str] = None,
+            get_remote_home_dir: Callable[[], str] = lambda: '~'):
```
Comment (cblmemo) [inline sky/utils/command_runner.py:227]:
```suggestion
            get_remote_home_dir: Callable[[], str] = lambda: '~') -> None:
```
nit; to enable mypy check

File: sky/utils/command_runner.py [hunk 5]
```diff
@@ -207,6 +212,92 @@ def _get_command_to_run(
...
+        if prefix_command is not None:
+            rsync_command.append(prefix_command)
+        rsync_command += ['rsync', RSYNC_DISPLAY_OPTION]
+
+        # --filter
+        rsync_command.append(RSYNC_FILTER_OPTION)
+
+        if up:
+            # The source is a local path, so we need to resolve it.
+            # --exclude-from
```
Comment (cblmemo) [inline sky/utils/command_runner.py:240]: What does this line mean?

File: sky/utils/command_runner.py [hunk 6]
```diff
@@ -207,6 +212,92 @@ def _get_command_to_run(
...
+                # Ensure file exists; otherwise, rsync will error out.
+                #
+                # We shlex.quote() because the path may contain spaces:
+                #   'my dir/.git/info/exclude'
+                # Without quoting rsync fails.
+                rsync_command.append(
+                    RSYNC_EXCLUDE_OPTION.format(
+                        shlex.quote(str(resolved_source / GIT_EXCLUDE))))
+
+        rsync_command.append(f'-e "{rsh_option}"')
```
Comment (cblmemo) [inline sky/utils/command_runner.py:252]:
```suggestion
        rsync_command.append(f'-e {shlex.quote(rsh_option)}')
```
to make it more robust?

File: sky/utils/command_runner.py [hunk 7]
```diff
@@ -584,3 +584,249 @@ def rsync(
                                            error_msg,
                                            stderr=stdout + stderr,
                                            stream_logs=stream_logs)
+
+
+class KubernetesCommandRunner(CommandRunner):
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/utils/command_runner.py:623]: Good point! Agree. Lets keep the current name ;)

File: sky/utils/command_runner.py [hunk 8]
```diff
@@ -584,3 +584,249 @@ def rsync(
...
+    ):
+        """Initialize KubernetesCommandRunner.
+
+        Example Usage:
+            runner = KubernetesCommandRunner((namespace, pod_name))
+            runner.run('ls -l')
+            runner.rsync(source, target, up=True)
+
+        Args:
+            node: The namespace and pod_name of the remote machine.
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/utils/command_runner.py:639]: Good point! Agree.

**Review by cblmemo — APPROVED** [review]
Thanks for the fix! LGTM except for several nits ;)

File: sky/utils/command_runner.py (hunk 5 above)
Comment (Michaelvll) [reply sky/utils/command_runner.py:240]: Updated the comment. thansk!

**Comment by Michaelvll** [comment]
This PR should be ready to go, cc'ing @romilbhardwaj for a final check before we get this in, as it may affect the kubernetes code path significantly

File: sky/backends/cloud_vm_ray_backend.py [hunk 9]
```diff
@@ -3644,7 +3644,10 @@ def _rsync_down(args) -> None:
             try:
                 os.makedirs(local_log_dir, exist_ok=True)
                 runner.rsync(
-                    source=f'{remote_log_dir}/*',
+                    # Require a `/` at the end to make sure the parent dir
+                    # are not created locally. We do not add additional '*' as
+                    # kubernetes's rsync does not work with an ending '*'.
+                    source=f'{remote_log_dir}/',
```
Comment (romilbhardwaj) [inline sky/backends/cloud_vm_ray_backend.py:3650]: Curious -  does the removal of `*` impact how hidden files are handled? On my mac it does not make any difference and this change should be ok, but [this article](https://www.baeldung.com/linux/rsync-and-hidden-files) seems to suggest it does. Any thoughts?

**Review by romilbhardwaj — APPROVED** [review]
Thanks @Michaelvll! Took a quick look at the code and tried it out, provisioning feels much snappier now 🚀 LGTM if kubernetes smoke tests pass.

File: sky/utils/command_runner.py [hunk 10]
```diff
@@ -542,43 +609,199 @@ def rsync(
...
+        command_str = self._get_command_to_run(cmd,
+                                               process_stream,
+                                               separate_stderr,
+                                               skip_lines=skip_lines,
+                                               source_bashrc=source_bashrc)
+        command = kubectl_base_command + [
+            # It is important to use /bin/bash -c here to make sure we quote the
+            # command to be run properly. Otherwise, directly appending commands
+            # after '--' will not work for some commands, such as '&&', '>' etc.
+            '/bin/bash',
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/utils/command_runner.py:722]: Good point! We have a bunch of places using `bash` and `bashrc`, assuming the existence of `bash` (even debian based image, due to the use of `apt install`). We should update those places as well to support those base images. : )

File: sky/backends/cloud_vm_ray_backend.py (hunk 9 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:3650]: It seems no significant difference between `/*` and `/`, and the latter is more robust. The tailing `*` means the shell expands the pattern to include all files and directories in src before rsync runs, while the latter relies on rsync's own logic to sync the content of src.

**Comment by Michaelvll** [comment]
Tested with `pytest tests/test_smoke.py::test_minimal --kubernetes` and passed. Merging now : )

**Pull request merged**